primitives/src/side_effect/mod.rs

```diff
@@ -86,22 +95,177 @@ impl Default for SecurityLvl {
+            Ok(confirmed.clone())
+        } else {
+            Err("At harden() expect FSX confirmation part to be there")
+        }?;
+
+        let confirmation_outcome = if let Some(outcome) = &confirmed.err {
+            outcome.clone()
+        } else {
+            ConfirmationOutcome::Success
+        };
+
+        let confirmed_cost: CircuitBalance = if let Some(cost) = &confirmed.cost {
+            Decode::decode(&mut &cost.encode()[..])
+                .map_err(|_| "harden() decoding error: confirmed_cost")
+        } else {
+            Ok(0u128)
+        }?;
+
+        let confirmed_executioner: AccountId32 =
+            Decode::decode(&mut &confirmed.executioner.encode()[..])
+                .map_err(|_| "harden() decoding error: confirmed_executioner")?;
+
+        let confirmed_received_at: CircuitBlockNumber =
+            Decode::decode(&mut &confirmed.received_at.encode()[..])
+                .map_err(|_| "harden() decoding error: confirmed_received_at")?;
+
+        let prize: CircuitBalance = Decode::decode(&mut &self.input.prize.encode()[..])
+            .map_err(|_| "harden() decoding error: confirmed_cost")?;
+
+        return Ok((
+            self.security_lvl.clone(),
+            confirmation_outcome,
+            confirmed_cost,
+            confirmed_executioner,
+            confirmed_received_at,
+            prize,
+            self.input.target.clone(),
+            self.input.encoded_args.clone(),
+            <[u8; 4]>::try_from(self.input.encoded_action.clone()).unwrap_or_default(),
+        ))
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct HardenedSideEffect {
+    target: [u8; 4],
+    prize: CircuitBalance,
+    encoded_action: [u8; 4],
+    encoded_args: Vec<Bytes>,
+    encoded_args_abi: Vec<crate::abi::Type>,
+
+    security_lvl: SecurityLvl,
+
+    confirmation_outcome: ConfirmationOutcome,
+    confirmed_executioner: AccountId,
+    confirmed_received_at: CircuitBlockNumber,
+    confirmed_cost: CircuitBalance,
+}
+
+impl Default for HardenedSideEffect {
+    fn default() -> Self {
+        HardenedSideEffect {
+            target: [0, 0, 0, 0],
+            prize: 0u128,
+            encoded_action: [0, 0, 0, 0],
+            encoded_args: vec![],
+            encoded_args_abi: vec![],
+            security_lvl: SecurityLvl::Dirty,
+            confirmation_outcome: ConfirmationOutcome::Success,
+            confirmed_executioner: AccountId32::new([0u8; 32]),
+            confirmed_received_at: 0,
+            confirmed_cost: 0,
+        }
+    }
+}
+
+impl
+    sp_runtime::traits::Convert<
+        (
+            SecurityLvl,
+            ConfirmationOutcome,
+            CircuitBalance,
+            AccountId32,
+            CircuitBlockNumber,
+            CircuitBalance,
+            [u8; 4],
+            Vec<Bytes>,
+            [u8; 4],
+        ),
+        HardenedSideEffect,
+    > for HardenedSideEffect
+{
+    fn convert(
```

Comment:
You can use the From<(..)> trait instead of custom convert

Reply:
So there are loads of implicit traits that usually imply From<..>. If we use From for all of our conversions we wont need to have any confusion. Generally whenever you need to transform some type to another you'd use From/TryFrom in rust.

Reply:
done!